File: SortAMatrix.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
// Hàm so sánh để sử dụng trong qsort
int compare(const void *a, const void *b) {return (*(int *)a - *(int *)b);}
// Hàm sắp xếp toàn bộ ma trận
void sapXepMaTran(int soHang, int soCot, int maTran[soHang][soCot]) {
    int tongSoPhanTu = soHang * soCot; // Tổng số phần tử trong ma trận
    int mangPhang[tongSoPhanTu];      // Mảng phụ để làm phẳng ma trận
    int chiSo = 0;                    // Chỉ số để duyệt qua các phần tử
    // Bước 1: Làm phẳng ma trận thành mảng 1 chiều
    for (int i = 0; i < soHang; i++) {
        for (int j = 0; j < soCot; j++) {
            mangPhang[chiSo++] = maTran[i][j];
        }
    }
    // Bước 2: Sắp xếp mảng 1 chiều bằng qsort
    qsort(mangPhang, tongSoPhanTu, sizeof(int), compare);
    // Bước 3: Ghi lại mảng đã sắp xếp vào ma trận
    chiSo = 0;
    for (int i = 0; i < soHang; i++) {
        for (int j = 0; j < soCot; j++) {
            maTran[i][j] = mangPhang[chiSo++];
        }
    }
}
// Hàm in ma trận
void inMaTran(int soHang, int soCot, int maTran[soHang][soCot]) {
    for (int i = 0; i < soHang; i++) {
        for (int j = 0; j < soCot; j++) {
            printf("%d ", maTran[i][j]);
        }
        printf("\n");
    }
}
int main() {
    int soHang, soCot;
    printf("Nhập số hàng của ma trận: "); scanf("%d", &soHang);
    printf("Nhập số cột của ma trận: "); scanf("%d", &soCot);
    int maTran[soHang][soCot];
    printf("Nhập các phần tử của ma trận, từng hàng một:\n");
    for (int i = 0; i < soHang; i++) {
        printf("Nhập hàng %d (gồm %d phần tử, cách nhau bởi dấu cách): ", i + 1, soCot);
        for (int j = 0; j < soCot; j++) {
            scanf("%d", &maTran[i][j]);
        }
    }
    // Sắp xếp toàn bộ ma trận
    sapXepMaTran(soHang, soCot, maTran);
    // In ma trận đã sắp xếp
    printf("Ma trận sau khi sắp xếp toàn bộ:\n");
    inMaTran(soHang, soCot, maTran);
    return 0;
}

File: SortAMatrix.py
# Hàm để sắp xếp toàn bộ ma trận
def sap_xep_ma_tran(ma_tran):
    # Bước 1: Làm phẳng ma trận thành một danh sách
    mang_phang = [phan_tu for hang in ma_tran for phan_tu in hang]

    # Bước 2: Sắp xếp danh sách đã làm phẳng
    mang_phang.sort()

    # Bước 3: Ghi lại giá trị từ danh sách đã sắp xếp vào ma trận
    chi_so = 0
    so_hang = len(ma_tran)
    so_cot = len(ma_tran[0]) if so_hang > 0 else 0
    for i in range(so_hang):
        for j in range(so_cot):
            ma_tran[i][j] = mang_phang[chi_so]
            chi_so += 1

# Hàm để in ma trận
def in_ma_tran(ma_tran):
    for hang in ma_tran:
        print(" ".join(map(str, hang)))

# Hàm chính
def main():
    # Nhập số hàng và cột
    so_hang = int(input("Nhập số hàng của ma trận: "))
    so_cot = int(input("Nhập số cột của ma trận: "))

    # Nhập từng hàng của ma trận
    ma_tran = []
    print("Nhập các phần tử của ma trận, từng hàng một:")
    for i in range(so_hang):
        hang = list(map(int, input(f"Nhập hàng {i + 1} (gồm {so_cot} phần tử, cách nhau bởi dấu cách): ").split()))
        while len(hang) != so_cot:  # Kiểm tra đúng số phần tử trên mỗi hàng
            print(f"Vui lòng nhập đúng {so_cot} phần tử!")
            hang = list(map(int, input(f"Nhập lại hàng {i + 1}: ").split()))
        ma_tran.append(hang)

    # Gọi hàm để sắp xếp toàn bộ ma trận
    sap_xep_ma_tran(ma_tran)

    # In ma trận sau khi sắp xếp
    print("Ma trận sau khi sắp xếp toàn bộ:")
    in_ma_tran(ma_tran)

# Chạy chương trình
if __name__ == "__main__":
    main()
